priv/posts/security/20241212210148_epmd-public-exposure.md

```diff
@@ -0,0 +1,114 @@
+}
+---
+EPMD, essential for Erlang and RabbitMQ clustering, is often exposed online—posing hidden security risks and requiring quick mitigation steps.
+---
+
+The Erlang Port Mapper Daemon (EPMD) is a built-in component that facilitates clustering across Erlang-based applications (including RabbitMQ). While EPMD is essential in trusted network environments, it was never intended for exposure on the public internet. Recent scans have revealed over 85,000 instances of publicly exposed EPMD, with roughly half associated with RabbitMQ servers.
+
+If left unsecured, an exposed EPMD endpoint can potentially grant attackers a foothold, enabling them to join your cluster, run arbitrary code, and compromise your systems. Fortunately, mitigation steps are straightforward: disable EPMD if you’re not clustering, or restrict it behind a firewall and proper network configuration.
+
+## TL;DR
+
+**Check if you’re exposed:**
+- Run a port scan:  
+  ```bash
+  nmap -p 4369 -sV [Your Public IP]
+  ```  
+  If the output shows:  
+  ```
+  PORT     STATE SERVICE
+  4369/tcp open  epmd
+  ```
+  Your EPMD port is exposed.
+
+**For Erlang/Elixir users:**
+- Remove or avoid using `-name` or `-sname` if you don’t need clustering.
+- Bind distribution to a non-public interface with `vm.args`:
+  ```bash
+  -kernel inet_dist_use_interface '{127, 0, 0, 1}'
+  ```
+- Use a firewall or security group rules to restrict external access.
+
+**For RabbitMQ users:**
+- RabbitMQ runs on Erlang and uses EPMD behind the scenes. Even if you never interact directly with Erlang, you may still be affected.
+- Check the RabbitMQ networking documentation to learn how to identify exposed ports and close them:  
+  [https://www.rabbitmq.com/docs/networking#epmd](https://www.rabbitmq.com/docs/networking#epmd)  
+- The Erlang Ecosystem Foundation’s security hardening guide explains what EPMD is, which ports it uses, and how to close them down:  
+  [https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/distribution](https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/distribution)
+
+## What is EPMD and How Does it Work?
+
+Erlang was designed from the ground up for concurrency, distribution, and fault tolerance. To support clustering (distributed Erlang nodes connecting and communicating seamlessly), Erlang relies on EPMD. EPMD runs as a small daemon that keeps track of Erlang nodes and their associated ports. When a node starts with `-name` or `-sname`, it registers with EPMD on the host machine’s port 4369.
+
+In a controlled environment—such as an internal network—this works well. EPMD makes it easy for nodes to discover each other, connect, and form distributed clusters.
+
+**Key points:**
+- EPMD listens on port 4369 by default.
+- It’s assumed that EPMD is only reachable from within a trusted network.
+- Many Erlang/Elixir build tools and release pipelines (like Rebar3, Mix, and Distillery releases) enable EPMD by default.
```

Comment:
Agreed

I see two possibilities about the defaults:

* Erlang by default only starts listening on localhost. Users will then either have to specify `inet_dist_use_interface` with `0.0.0.0` for all interfaces or list specific ones.
* The tools by default don't include `-name` / `-sname` or include ` -kernel inet_dist_use_interface '{127, 0, 0, 1}'`

I personally would prefer to change the default interfaces in Erlang, which will always require a conscious decision by the user.

Also: Not sure if the Mix default is actually problematic with the `remote.vm.args`. I'm not sure how that actually works and will need to investigate. https://github.com/erlef/website/pull/493#discussion_r1882937503

Reply:
> I personally would prefer to change the default interfaces in Erlang, which will always require a conscious decision by the user.

By doing that change, you will break every guide/book/video that exists today showing how to setup Erlang Distribution and also all Erlang systems that use distribution correctly. All for a security issue that really should be solved by fixing the firewall to be secure. It would also not fix the RabbitMQ/ejabbard nodes, as the vast majority of such deployments would need Erlang Distribution and EPMD for fault-tolerance.

Also I would again like to point out that Erlang is safe by default, as it does not enable distribution. It is a concious decision by tools and users to enable it. If you enable something called Distributed Erlang, can you really expect that to not listen to external interfaces?

IMO relx, mix release and others should either limit both epmd and dist interface to localhost by default, or just disable it completely and run an [Erlang ssh daemon](https://www.erlang.org/doc/apps/ssh/using_ssh.html#running-an-erlang-ssh-daemon) as part of the release that you can connect to for debugging.